### ISSUE TITLE
Automatic binning

### ISSUE THREAD

**Issue by MuellerSeb** [GeoStat-Framework/GSTools#55]
This is just a note, so we don't forget and can talk about it:

My proposal for default bining:

- number of bins calculated by either:
  - (a) Sturges rule:

        n_bins = 2 * np.log2(n_points) + 1

  - (b) Rice rule:

        n_bins = np.power(n_points, 2 / 3)

- maximal bin edge as a third of the box diameter:

      diam = np.sqrt((x.max - x.min)**2 + (y.max - y.min)**2 + (z.max - z.min)**2)
      bin_max = diam / 3

- resulting bin sizes (quadratic growth):

      bins = np.linspace(0, bin_max**(1/2), n_bins)**2

Example: 30 points with ``diam = 900`` and Struges rule:

    bins = np.linspace(0, (900/3)**0.5, int(np.around(2*np.log2(30) + 1))) ** 2
    array([  0.,   3.,  12.,  27.,  48.,  75., 108., 147., 192., 243., 300.])

**Comment by banesullivan** [comment]
@MuellerSeb, what is the `d` in 

```
diam = ((x.max - x.min)**d + (y.max - y.min)**d + (z.max - z.min)**d)**(1/d)
```

**Comment by MuellerSeb** [comment]
The dimension... sorry

**Comment by MuellerSeb** [comment]
It should be contantly 2, since it is the euclidean norm.. I just mixed it up with the d-norm. 

Comment was updated.

**Comment by banesullivan** [comment]
@MuellerSeb, I did this for the example on #36 and it has really great results. The variogram doesn't look all that pretty, but the results were compelling.

I will make a PR with this now

**Comment by MuellerSeb** [comment]
Thanks for your PR. @LSchueler is working on this at the moment, so I have closed your PR in favor of the re-written routines.

But it is good to hear, that the scheme was working for you! That means we are on the right way.

**Comment by MuellerSeb** [comment]
@LSchueler: It would be nice to get some information on the variance within each bin during the variogram estimation. One could use a sequential update during the estimation, so we don't consume memory:

![mean](https://wikimedia.org/api/rest_v1/media/math/render/svg/72192362c590b89684309e78f6644ab0864f8d1b)
![var](https://wikimedia.org/api/rest_v1/media/math/render/svg/a07187029f774ee0da058f3b6ab6e5d3e7131df4)

We could later use this variance as a target function to estimate anisotropy and rotation by minimizing it. Or we could use this as weights during the fitting of a variogram function.

**Comment by MuellerSeb** [comment]
Then we could get the following values for each bin:
* bin center
* estimate variogram at the bin center
* count of values within the bin
* mean of values within the bin
* variance of values within the bin

**Comment by TobiasGlaubach** [comment]
Hi everyone,
first of all thanks for pointing me here @MuellerSeb.
Did you try the sequential update scheme before somewhere @MuellerSeb? I tried it for a realtime sensordata processing pipeline a few years ago and I found that the numerical error accumulates much much quicker, compared with the standard approach. I am not sure however on the datatypes and data I used back then. Just saying that this should be something to test, when implementing.

**Comment by TobiasGlaubach** [comment]
Regarding your idea [here](https://github.com/GeoStat-Framework/GSTools/issues/83#issuecomment-620548675) I think it's a good idea and I tried something similar before (the rotation and stretching part not the minimizing variance part) what I failed to accomplish (in my quick tests without investing much time) was to properly de-stretch the data, since one would need to stretch based on variance. My problem was, that very anisotrophic and very sparse data (well drilling data) let to bad estimation results. I did not invest a lot of time in it though.

Regarding the rotation:
Scipy has a fast and efficient implementation for this.

Generating a test model:

```python
from scipy.spatial.transform import Rotation as R
import numpy as np
import gstools as gs
import matplotlib.pyplot as plt
```

```python

x = y = np.arange(100)
model = gs.Exponential(dim=2, var=1, len_scale=[12.0, 3.0], angles=np.pi / 8)
srf = gs.SRF(model, seed=20170519)
srf.structured([x, y])
srf.plot()
val = srf.field.copy()
(gridx, gridy) = srf.pos
```
![image](https://user-images.githubusercontent.com/34371519/80491721-648dbe80-8963-11ea-8dad-94f4701f73f5.png)

Selecting a random unstructured subset of the data:

```python

X, Y = np.meshgrid(gridx, gridy)
idx = np.random.choice(X.size, 5000)
x = X.flatten()[idx]
y = Y.flatten()[idx]
v = val.flatten()[idx]
```

And rotating it:

```python
r = R.from_euler('z', 180-67, degrees=True)
xyz = np.stack((x, y, np.zeros_like(x))).T

xyzr = np.apply_along_axis(r.apply, 1, xyz)
f, axes = plt.subplots(1,2, figsize=(10,5))
axes[0].scatter(xyz[:,0], xyz[:,1], c=v, marker='.')
axes[1].scatter(xyzr[:,0], xyzr[:,1], c=v, marker='.')
```
  
![image](https://user-images.githubusercontent.com/34371519/80492056-ccdca000-8963-11ea-9571-5bfee893724a.png)


I found that for my examples the speed of the rotation operation by far fast enough.

**Comment by TobiasGlaubach** [comment]
> @LSchueler: It would be nice to get some information on the variance within each bin during the variogram estimation. One could use a sequential update during the estimation, so we don't consume memory:
> 
> ![mean](https://camo.githubusercontent.com/f27fd7c7521bed110f7785b36d4aafcc56cd91ed/68747470733a2f2f77696b696d656469612e6f72672f6170692f726573745f76312f6d656469612f6d6174682f72656e6465722f7376672f37323139323336326335393062383936383433303965373866363634346162303836346638643162)
> ![var](https://camo.githubusercontent.com/7e7db3ad31b7d541e13dd3ca9000cd455f231f7e/68747470733a2f2f77696b696d656469612e6f72672f6170692f726573745f76312f6d656469612f6d6174682f72656e6465722f7376672f61303731383730323966373734656530646130353866336236616236653564336537313331646634)
> 
> Where the `x_n` are the samples within one bin.
> 
> We could later use this variance as a target function to estimate anisotropy and rotation by minimizing it. Or we could use this as weights during the fitting of a variogram function.

I just did a quick test and found that there is a difference between the formulas you proposed and 
the mean and var implementations of numpy. Not sure If I did everything right and not sure if I am comparing apples to oranges here, but just putting this out for further discussions:

```python
import numpy as np

for n in [100, 1000, 10000, 100000]:
    
    x = np.random.randn(n,1) * 3 + 1
    ninv = 1./float(n)
    print((n, ninv, np.mean(x)))
    
    x_var_i = 0
    x_dash_i = 0
    for i in range(n):
        x_dash_i_new = x_dash_i + 1/n * (x[i] - x_dash_i)
        x_var_i = x_var_i + 1/n * ( (x[i] - x_dash_i) * (x[i] - x_dash_i_new) - x_var_i )
        x_dash_i = x_dash_i_new
                                   
    print(('mean:', np.mean(x), x_dash_i, np.abs(np.mean(x) - x_dash_i)))
    print(('var :', np.var(x), x_var_i, np.abs(np.var(x) - x_var_i)))
    print('')
```

yields:
```
(100, 0.01, 1.6196098015504952)
('mean:', 1.6196098015504952, array([1.02193216]), array([0.59767764]))
('var :', 8.07784019968546, array([5.53715413]), array([2.54068607]))

(1000, 0.001, 0.9247298725125604)
('mean:', 0.9247298725125604, array([0.58335667]), array([0.34137321]))
('var :', 8.79624573553866, array([5.78028449]), array([3.01596124]))

(10000, 0.0001, 0.9590076568417759)
('mean:', 0.9590076568417759, array([0.6073434]), array([0.35166426]))
('var :', 8.983778117447093, array([5.88507919]), array([3.09869892]))

(100000, 1e-05, 1.0068634559281715)
('mean:', 1.0068634559281715, array([0.63635949]), array([0.37050397]))
('var :', 9.031393103173771, array([5.93670523]), array([3.09468787]))

```

**Comment by MuellerSeb** [comment]
@TobiasGlaubach : Indeed they should match. The advantage of the sequential formulation is, that you can update the current value for variance, when there comes new data. Since the cython routines loop over all pairs of points and then decides for the bin to put these combination into, it is more convenient to use a sequential update of the variance value, so we don't have to store all values to calculate the variance afterwards (memory efficiency).

I guess the difference in the variance formulation comes from the use of population vs. sample variance: https://en.wikipedia.org/wiki/Variance#Population_variance_and_sample_variance

where the only difference is the divisor:
- `n` (population variance)
- `n-1` (sample variance)

For increasing `n` they are converging to each other as you observed correctly. You can specify this in the numpy routine by `ddof`: https://numpy.org/doc/stable/reference/generated/numpy.var.html

Aaaand in your code-sample, you are starting at `n=1` so you are keeping out the first value.

Regarding your de-stretching question, GSTools provides routines to de-stretch and re-rotate fields given here:
https://github.com/GeoStat-Framework/GSTools/blob/2c26d749ecd8554f10579bef4ac2d1f70266b5a9/gstools/field/tools.py#L114

https://github.com/GeoStat-Framework/GSTools/blob/2c26d749ecd8554f10579bef4ac2d1f70266b5a9/gstools/field/tools.py#L136

**Comment by TobiasGlaubach** [comment]
Thanks for looking into this. Actually the starting at 1 is no problem if one sets 

```python
    x_var_i = 0
    x_dash_i = x[0]
```

I am currently working on estimate.pyx for the directional variogram estimation in #83. I can implement the variance and mean estimation as well. So it can be used for other things like the mentioned anisotrophy and orientation fitting.
@MuellerSeb any reason against that?

**Comment by MuellerSeb** [comment]
Closed by #131
